src/server.js

```diff
@@ -0,0 +1,319 @@
+      
+      console.log("Trying " + url);
+
+      try 
+      {
+        //request the URL
+        const request = bent(url);
+        const response = await request();
+        //If OK
+        if (200 == response.statusCode)
+        {
+          console.log("HTTP/S services found at: " + url)
+          return url;
+        }
+      }
+      catch(error)
+      {
+        //TODO: Handle expected connection errors, but re-throw anything else
+      }
+    })).then((urls) =>
+    {
+      // An array of the promise resolves will be returned. The discovered services
+      // will have their URL in the array. The others will be null.
+      var filteredUrls = urls.filter(function (el) {
+        // Only keep the non-null items
+        return el != null;
+      });
+
+      // If we found a URL
+      if(filteredUrls.length > 0)
+      {
+        returnURL = filteredUrls[0];
+      }
+      // Otherwise send the default HTML
+      else
+      {
+        console.log("Displaying default HTML page");
+       returnURL = "file:///home/chromium/index.html";
+      }
+    });
+
+    return returnURL;
+  }
+       
+// Launch the browser with the URL specified
+let launchChromium = function(url) {
+  chromeLauncher.killAll().then(() => { 
+
+    flags = [];
+    // If the user has set the flags, use them
+    if (null != FLAGS)
+    {
+      flags = FLAGS.split(' ');
+    }
+    else
+    {
+      // User the default flags from chrome-launcher, plus our own.
+      flags = DEFAULT_FLAGS;
+      var balenaFlags = [
+        '--window-size=' + WINDOW_SIZE,
+        '--window-position=' + WINDOW_POSITION,
+        '--autoplay-policy=no-user-gesture-required',
+        '--noerrdialogs',
+        '--disable-session-crashed-bubble',
+        '--check-for-update-interval=31536000',
+        '--disable-dev-shm-usage' // TODO: work out if we can enable this for devices with >1Gb of memory
+      ];
+
+      // Merge the chromium default and balena default flags
+      flags = flags.concat(balenaFlags);
+    }
+
+    if (enableGpu != '1')
+    {
+      console.log("Disabling GPU")
+      flags.push('--disable-gpu')
+    }
+    else
+    {
+      console.log("Enabling GPU")
+    }
+
+    if (PERSISTENT_DATA == '1')
+    {
+      flags.push('--user-data-dir=/data')
+    }
+
+    var startingUrl = url;
+    if (kioskMode == '1')
+    {
+      console.log("Enabling KIOSK mode")
+      startingUrl = '--app=' + url
+    }
+    else
+    {
+      console.log("Disabling KIOSK mode")
+    }
+
+    console.log("Starting Chromium with the following flags: " + flags)
+    console.log("Displaying URL " + startingUrl)
+
+    chromeLauncher.launch({
+      startingUrl: startingUrl,
+      ignoreDefaultFlags: true,
+      chromeFlags: flags,
+      port: REMOTE_DEBUG_PORT,
+      userDataDir: '/tmp/balena'
+    }).then(chrome => {
+      console.log(`Chromium debugging port running on port: ${chrome.port}`);
+      currentUrl = url;
+    });
+  });
+}
```

Comment:
Again avoid using `.then`. If you can do `something.then()` it returns a promise, so you can also `await something()`

Reply:
Also, as it's coded now, the function `launchChromium` does not wait for the internal `.then()` calls to finish. It's probably not a problem in your case but it could. If you `await` the internal calls you can turn `launchChromium` into an `async` function.